_posts/2023-10-17-rust-in-yocto.md

```diff
@@ -0,0 +1,149 @@
+
+The progress on meta-rust was further discussed in a [November 2022 presentation](https://www.youtube.com/watch?v=7uCzL2ZwRMU), which highlighted the maintainers' current focus on enhancing testing, reproducibility, and minimizing build times.
+
+## Packaging a Rust Program with the cargo Class
+
+Regardless of whether one opts for the official Rust support or relies on `meta-rust/meta-rust`, the packaging steps for a Rust program remain consistent. (Obviously, if you're using a Yocto version without built-in Rust support or have an inclination towards a newer Rust version, you'd need to clone the `meta-rust/meta-rust` layer and include it in your `conf/layers.conf` file.)
+
+Consider the creation of a rudimentary Rust project with an added dependency:
+
+```shell
+$ cargo init .
+$ cat src/main.rs
+fn main() {
+    println!("Hello, world!");
+}
+$ cargo add serde
+```
+
+Though only one package was appended, it brings along its dependencies, reflected in the six dependencies now listed in `Cargo.lock`.
+
+The utility [`cargo-bitbake`](https://github.com/meta-rust/cargo-bitbake) comes in handy when formulating a recipe for our program. It peruses the `Cargo.toml` and `Cargo.lock` files, generating a Yocto recipe based on the provided information.
+
+```
+$ cargo install --locked cargo-bitbake
+$ cargo bitbake
+```
+
+The generated recipe includes all the essential elements required for a Yocto build, with a comprehensive listing of the dependencies and their respective versions. The essential `inherit cargo` line automates the process of building a Rust project with a `Cargo.toml` file, even accommodating the building of C dependencies if a `build.rs` exists.
+
+Embedding this recipe within an existing layer and initiating `bitbake hellorust` builds the rust cross compiler and our program is built for the target.
+
+## meta-rust-bin
+
+[`meta-rust-bin`](https://github.com/rust-embedded/meta-rust-bin/tree/master), a project of the rust-embedded group, emerges as a sought-after alternative for packaging Rust programs within Yocto. As an OpenEmbedded/Yocto layer, it uses official pre-built rust toolchains.
+
+Among its other merits are its frequent updates aligning with the latest Rust versions and its support for older Yocto versions, ranging [from dunfell to langdale](https://github.com/rust-embedded/meta-rust-bin/blob/ef9a765b0785ed9e964691ac1a5a302dd0a0ea62/conf/layer.conf#L13-L20).
+
+Previously, `meta-rust-bin` presented a `cargo` class, rendering it incompatible with `meta-rust`. The recommendation was to use a `BBMASK` statement to "hide" the built-in `cargo` and `rust` classes. Fortunately, this stumbling block was recently addressed with [the classname undergoing a change to `cargo_bin`](https://github.com/rust-embedded/meta-rust-bin/pull/136), enabling seamless usage of both the official Rust support and meta-rust-bin within the same Yocto project.
```

Comment:
Nice, I hadn't seen that this got merged.

Reply:
Let's crosslink to previous webinars to boost SEO rankings. Recommend placing this before the subscription and table of contents callouts. Thomas, up to you if it should be phrased in the third person (as it's currently structured), or if you'd like to change it to "Check out MY previous webinars..."

> 📺 **Watch the Webinar Recordings**
>
> Check out Thomas' previous webinars for more Embedded Linux discussions: [Over-the-Air Updates for Embedded Linux Devices](https://go.memfault.com/over-the-air-updates-embedded-linux-devices) and [Wrangling Penguins: Better Embedded Linux Monitoring and Debugging](https://go.memfault.com/better-embedded-linux-monitoring-debugging-memfault).

Reply:
added - I think the 3rd person makes it a little less "sales-y" so I kept it.